include/drivers/can.h

```diff
@@ -325,14 +334,95 @@ struct can_driver_api {
 
 };
 
+/**
+ * @brief Convert the DLC to the number of bytes
+ *
+ * This function converts a the Data Length Code to the number of bytes.
+ *
+ * @param dlc The Data Length Code
+ *
+ * @retval Number of bytes
+ */
+
+static inline u8_t can_dlc_to_bytes(u8_t dlc)
+{
+	static const u8_t dlc_table[] = {1, 2, 3, 4, 5, 6, 7, 8, 12,
+					 16, 20, 24, 32, 48, 64};
+
+	return dlc > 0x0F ? 64 : dlc_table[dlc];
+}
+
+/**
+ * @brief Convert a number of bytes to the DLC
+ *
+ * This function converts a number of bytes to the Data Length Code
+ *
+ * @param num_bytes The number of bytes
+ *
+ * @retval The DLC
+ */
+
+static inline u8_t can_bytes_to_dlc(u8_t num_bytes)
+{
+	return num_bytes <= 8  ? num_bytes :
+	       num_bytes <= 12 ? 9 :
+	       num_bytes <= 16 ? 10 :
+	       num_bytes <= 20 ? 11 :
+	       num_bytes <= 24 ? 12 :
+	       num_bytes <= 32 ? 13 :
+	       num_bytes <= 48 ? 14 :
+	       15;
+}
+
+#ifdef CONFIG_CAN_FD_MODE
+
+/**
+ * @brief Perform data transfer to CAN bus with flexible datarate option.
+ *
+ * This routine provides a generic interface to perform data transfer
+ * to the can bus with options for CAN-FD
+ * *
+ * @param dev          Pointer to the device structure for the driver instance.
+ * @param frame        Frame to transfer.
+ * @param brs          Baudrate Switch. If true, switch to fast br in dataphase.
+ * @param fd_format    Send the frame in CAN-FD frame format. Allows dlc > 8.
+ * @param timeout      Waiting for empty tx mailbox timeout in ms or K_FOREVER.
+ * @param callback_isr Is called when message was sent or a transmission error
+ *                     occurred. If NULL, this function is blocking until
+ *                     message is sent. This must be NULL if called from user
+ *                     mode.
+ * @param callback_arg This will be passed whenever the isr is called.
+ *
+ * @retval 0 If successful.
+ * @retval CAN_TX_* on failure.
+ */
+__syscall int can_fd_send(struct device *dev, const struct zcan_frame *frame,
+		       bool brs, bool fd_format, s32_t timeout,
```

Comment:
Unsure if there is a guideline about this, but this is becoming quite a long list of parameters. Maybe the bools can be merged into a flag bitmask?

Reply:
Maybe we can add this flags to the zcan_frame struct. I‘ll check that next week. Maybe we have some space thats padding so far.

Reply:
With the changes from https://github.com/alexanderwachter/zephyr/blob/can_rework_structs/include/drivers/can.h we could use the first reserved bit to distinguish fd frames.